File: C++/DFS/DFS (adjacency list).cpp
#include<iostream>
#include<queue>
#include<vector>
using namespace std;

bool visited[15];
vector <int> adj[15];
queue<int> ans;

void adj_check()
{
    // adj[1][2,6,7,9]
    adj[1].push_back(2);
    adj[1].push_back(6);
    adj[1].push_back(7);
    adj[1].push_back(9);
    
    //adj[2][3,4]
    adj[2].push_back(3);
    adj[2].push_back(4);
    
    //adj[4][5]
    adj[4].push_back(5);
    
    //adj[7][8]
    adj[7].push_back(8);
}

void dfs(int cur)
{
    visited[cur] = true;
    ans.push(cur);
    
    for(auto elem: adj[cur])
    {
        if(!visited[elem])
            dfs(elem);
    }
}

void ans_print()
{
    cout << "search order\n";
    cout << ans.front(); ans.pop();
    while(!ans.empty())
    {
        int cur = ans.front(); ans.pop();
        cout << "->"<<cur;
    }
    cout <<"\n";
}
           
int main()
{
    adj_check();
    dfs(1);
    ans_print();
    exit(0);
}

File: Python/DFS/DFS (adjacency list).py
from collections import deque


# adjacency list
def adj_check(n, m):
    global adj
    if n not in adj.keys():
        adj[n] = []

    adj[n].append(m)


def dfs(cur):
    global adj, visited, ans

    visited[cur] = True
    ans.append(cur)

    # not linked to other node
    if cur not in adj.keys():
        return

    for elem in adj[cur]:
        if not visited[elem]:
            dfs(elem)


def ans_print():
    global ans
    print("dfs search order")
    temp = ans.popleft()
    print(temp, end="")

    while len(ans) > 0:
        print("->{}".format(ans.popleft()), end="")


if __name__ == '__main__':
    visited = [False for _ in range(15)]
    adj = {}
    ans = deque()

    # 1: [2,6,7,9]
    adj_check(1, 2)
    adj_check(1, 6)
    adj_check(1, 7)
    adj_check(1, 9)

    # 2: [3,4]
    adj_check(2, 3)
    adj_check(2, 4)

    # 4: [5]
    adj_check(4, 5)

    # 7: [8]
    adj_check(7, 8)

    dfs(1)
    ans_print()